### PR TITLE
feat(entity): add 'setOne' method

### LINKED ISSUE #2369: @ngrx/entity add setOne (replaceOne?) method
Currently, the only way to replace an item is by using upsertOne. That does not work well with 'undefined' though (which is the case when the backend omits nulls when serializing to JSON). For example, i have an entity in the collection: {id: 1, name: 'foo', note: 'lorem ipsum'}. I want to replace it with {id: 1, name: 'foo'}. Obviously, upsertOne does not do the job well. Since the object  is already there, it attempts to perform an update, thus treating the new object as a patch and leaving 'note' value unchanged. So, a new method 'setOne' is needed here.

## Describe any alternatives/workarounds you're currently using

removeOne/addOne does the trick, but is not very efficient.

## If accepted, I would be willing to submit a PR for this feature

[X ] Yes (Assistance is provided if you need help submitting a pull request)
[ ] No

### PR DESCRIPTION
**Pull request by dmytro-gokun**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [X] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[] Bugfix
[X] Feature
[] Code style update (formatting, local variables)
[] Refactoring (no functional changes, no api changes)
[] Build related changes
[] CI related changes
[] Documentation content changes
[] Other... Please describe:
```

## What is the current behavior?

Closes #2369

## Does this PR introduce a breaking change?

```
[] Yes
[X] No
```

### CODE REVIEW

File: modules/entity/spec/sorted_state_adapter.spec.ts [hunk 1]
```diff
@@ -407,4 +407,36 @@ describe('Sorted State Adapter', () => {
...
+
+  it('should let you replace an entity in the state with setOne()', () => {
+    const withOne = adapter.addOne(
+      TheGreatGatsby,
+      adapter.addOne(AnimalFarm, adapter.addOne(AClockworkOrange, state))
+    );
+    const updatedBook = {
+      id: TheGreatGatsby.id,
+      title: 'A New Hope',
+      description: undefined,
```
Comment (alex-okrushko) [inline modules/entity/spec/sorted_state_adapter.spec.ts]: Could you please remove the description altogether? With `description: undefined` even `upsertOne` would behave the same way.

Comment (dmytro-gokun) [reply modules/entity/spec/sorted_state_adapter.spec.ts]: I can remove it, that is not a problem. But that is the only way to test if 'setOne' actually behaves in a way different from 'upsertOne'. I mean, without this 'undefined' the test does not make much sense. Does it?

File: modules/entity/spec/sorted_state_adapter.spec.ts [hunk 2]
```diff
@@ -407,4 +407,36 @@ describe('Sorted State Adapter', () => {
...
+      entities: {
+        [TheGreatGatsby.id]: TheGreatGatsby,
+      },
+    });
+  });
+
+  it('should let you replace an entity in the state with setOne()', () => {
+    const withOne = adapter.addOne(
+      TheGreatGatsby,
+      adapter.addOne(AnimalFarm, adapter.addOne(AClockworkOrange, state))
```
Comment (timdeschryver) [inline modules/entity/spec/sorted_state_adapter.spec.ts]:
```suggestion
      state
```

File: docs/entity/adapter.md [hunk 3]
```diff
@@ -87,6 +87,7 @@ state if no changes were made.
 - `addMany`: Add multiple entities to the collection
 - ~~`addAll`~~: (Deprecated and renamed to `setAll`). ~~Replace current collection with provided collection~~
 - `setAll`: Replace current collection with provided collection
+- `setOne`: Add or Replace on entity in the collection
```
Comment (timdeschryver) [inline docs/entity/adapter.md]: Could you move these changes to https://github.com/ngrx/platform/blob/master/projects/ngrx.io/content/guide/entity/adapter.md so they become visible on the website 🙂?
These docs are for older versions of NgRx

File: modules/entity/spec/sorted_state_adapter.spec.ts (hunk 2 above)
Comment (dmytro-gokun) [reply modules/entity/spec/sorted_state_adapter.spec.ts]: I'm not sure if I understand this. The reason I'm adding three books to the collection here is to test that sorted adapter preserves the correct sorting order after a 'setOne' call. If I only have one book in the collection that is not checked anymore. Do you think that is okay?

File: docs/entity/adapter.md (hunk 3 above)
Comment (dmytro-gokun) [reply docs/entity/adapter.md]: Done

File: modules/entity/spec/sorted_state_adapter.spec.ts (hunk 2 above)
Comment (timdeschryver) [reply modules/entity/spec/sorted_state_adapter.spec.ts]: Oh yea, that makes sense! Perhaps rename the `withOne` variable then tho

Comment (dmytro-gokun) [reply modules/entity/spec/sorted_state_adapter.spec.ts]: Oops. That's a copy&paste error :(. Should be fixed now.

File: modules/entity/spec/sorted_state_adapter.spec.ts (hunk 1 above)
Comment (alex-okrushko) [reply modules/entity/spec/sorted_state_adapter.spec.ts]: It actually does make more sense without that property. You can make a comment to be very explicit that we are updating the the entity that doesn't have that property and `toEqual` does the deep equality check.
With the test as-is even `adapter.upsertOne(updatedBook, withOne)` would pass.

I see that this PR is merge, I'd suggest to have a follow-up PR to adjust this test behavoir.

Comment (dmytro-gokun) [reply modules/entity/spec/sorted_state_adapter.spec.ts]: Okay, we might have a misunderstanding here... Do you think we need to:
1) remove line 429;
or
2) remove 'description' property from the Book interface altogether?

Comment (alex-okrushko) [reply modules/entity/spec/sorted_state_adapter.spec.ts]: remove the line from 429 and replace with `/* description property is not provided */`

**Pull request merged**